Enable `c++` source coverage with `CircleCI` and `codecov.io`

scintiller opened the pull request:
Summary:
Based on D23351877 (https://github.com/pytorch/pytorch/commit/1bda5e480c0fe01dae8db718cd457befa1b46ac6), life becomes much easier for this time.
1.in `build.sh`
- Enable coverage builld option for c++
- `apt-get install lcov`

2.in `test.sh`
- run `lcov`

3.in `pytorch-job-specs.yml`
- copy coverage.info to `test/` folder and upload it to codecov.io

Test Plan: Test on github

Differential Revision: D23464656

malfet commented on .circleci/verbatim-sources/job-specs/pytorch-job-specs.yml:
```diff
@@ -206,8 +206,10 @@ jobs:
               docker cp $id:/var/lib/jenkins/workspace/test/coverage.xml ./test
+              docker cp $id:/var/lib/jenkins/workspace/build/coverage.info ./test
               python3 -mpip install codecov
               python3 -mcodecov
+              python3 -mcodecov -f coverage.info
```
This will upload the report twice, wouldn't it?

scintiller replied:
Yes, I think it will. Fixed

malfet commented on .jenkins/pytorch/build.sh:
```diff
@@ -310,3 +310,13 @@ if [[ "${BUILD_ENVIRONMENT}" == *xla* ]]; then
   assert_git_not_dirty
 fi
+
+if [[ "$BUILD_ENVIRONMENT" ==  *coverage* ]]; then
```
This condition would never be executed, as there are no build job which name contains the word `coverage` defined in https://github.com/pytorch/pytorch/blob/master/.circleci/cimodel/data/pytorch_build_definitions.py

scintiller replied:
I fixed it by adding a new `build` job: `pytorch_linux_bionic_py3_8_gcc9_coverage_build` in `config.yml`. Not sure if it is the correct way.

seemethere replied:
That should work, but adding a new build *just* for coverage seems a bit overkill.

Is there a way to piggy-back this onto a separate build so we don't have to add a completely new configuration?

scintiller replied:
Yeah, it takes 2.5 hours to run the new build. Now I revise the existing `pytorch_linux_bionic_py3_8_gcc9_build` to `pytorch_linux_bionic_py3_8_gcc9_coverage_build`. Does this make sense?

seemethere commented on .jenkins/pytorch/test.sh:
```diff
@@ -14,6 +14,8 @@ echo "Testing pytorch"
   # TODO move this to docker
   pip_install unittest-xml-reporting coverage
+  sudo apt-get -qq update
+  sudo apt-get -qq install lcov
```
Is it possible to just add this to the base dockerfile instead of installing it in this test script?

scintiller replied:
Yes, this is better. Do you know where is the base dockerfile config?

scintiller replied:
Besides, I also wonder why we don't put it under `if [[ "$BUILD_ENVIRONMENT" == *coverage* ]]; then...` at line 36

seemethere commented on .jenkins/pytorch/test.sh:
```diff
@@ -401,9 +403,18 @@ else
     popd
+
+    # code coverage for c++ sources
+    pushd build
```
This is causing your `lcov ... build/` to fail since you are in the build directory when you try to run it

scintiller replied:
Fixed

malfet commented on .jenkins/pytorch/build.sh:
```diff
@@ -199,6 +199,13 @@ if [[ "${BUILD_ENVIRONMENT}" == *xla* ]]; then
+if [[ "$BUILD_ENVIRONMENT" ==  *coverage* ]]; then
+  # enable build option in CMake
+  export USE_CPP_CODE_COVERAGE=ON
+  export BUILD_TEST=ON
+  export CMAKE_BUILD_CONFIG=Debug
```
Those two should not be needed, should they?

malfet commented:
Please modify https://github.com/pytorch/pytorch/blob/master/codecov.yml to map `.AVX2.cpp` files back to their sources

malfet commented on .jenkins/pytorch/test.sh:
```diff
@@ -401,9 +412,18 @@ else
+    echo "Generating lcov coverage report for c++ sources"
+    lcov --capture --directory . --output-file coverage.info
+    # remove files that we do not care about from coverage report
+    lcov --remove coverage.info '/usr/*' '*third_party/*' '*build/*' --output-file coverage.info
```
We should not exclude build folder here, as it contains all native cpu kernels
```suggestion
    lcov --remove coverage.info '/usr/*' '*third_party/*'  --output-file coverage.info
```

malfet replied:
Same as above, please report time it takes to finish this command
```suggestion
    time lcov --remove coverage.info '/usr/*' '*third_party/*' --output-file coverage.info
```

malfet commented on .jenkins/pytorch/test.sh:
```diff
@@ -401,9 +412,18 @@ else
+    # code coverage for c++ sources
+    pushd build
+    echo "Generating lcov coverage report for c++ sources"
+    lcov --capture --directory . --output-file coverage.info
```
Please add `time` to report how long it takes to run this command
```suggestion
    time lcov --capture --directory . --output-file coverage.info
```

seemethere commented on .jenkins/pytorch/test.sh:
```diff
@@ -14,6 +14,17 @@ echo "Testing pytorch"
 if [ -n "${IN_CIRCLECI}" ]; then
   # TODO move this to docker
   pip_install unittest-xml-reporting coverage
+  # install lcov with 1.15 (gcc 9 requires lcov>1.14-2) outside of workplace, so we don't need to clean it up
```
If we only need this for code coverage would it make sense to move this to the `"${BUILD_ENVIRONMENT}" == *coverage*` block down below?

We probably don't need to be installing this for every single test job we run.

seemethere commented on .jenkins/pytorch/test.sh:
```diff
@@ -14,6 +14,17 @@ echo "Testing pytorch"
+  popd
+  popd
+  # symbolic link: make gcov-9 as default version of gcov, because gcc's version is gcc 9
+  sudo rm -r /usr/bin/gcov
```
Same goes for here, if we only need this for code coverage we should probably only do this in the `coverage` block we have below

seemethere reviewed (commented):
Looks mostly good to me, just a few minor changes to be made.

malfet commented on .jenkins/pytorch/test.sh:
```diff
@@ -12,9 +12,6 @@ source "$(dirname "${BASH_SOURCE[0]}")/common.sh"
 echo "Testing pytorch"
 
 if [ -n "${IN_CIRCLECI}" ]; then
-  # TODO move this to docker
-  pip_install unittest-xml-reporting coverage
-
```
Please leave `pip_install unittest-xml-reporting` in the default clause (otherwise test status reporting is lost of all but code coverage linux config)

malfet commented on .jenkins/pytorch/test.sh:
```diff
@@ -33,6 +30,20 @@ if [ -n "${IN_CIRCLECI}" ]; then
+    pip_install unittest-xml-reporting coverage
+    # install lcov with 1.15 (gcc 9 requires lcov>1.14-2, apt-get will only install 1.12)
+    # outside of workplace, so we don't need to clean it up
+    pushd ..
+    wget https://github.com/linux-test-project/lcov/releases/download/v1.15/lcov-1.15.tar.gz
+    tar -xf lcov-1.15.tar.gz
+    pushd lcov-1.15
+    sudo make install
+    popd
+    popd
```
Since lcov is just the Perl script, wouldn't it be easier to just download https://raw.githubusercontent.com/linux-test-project/lcov/master/bin/lcov into `~/local/bin`

malfet commented on .jenkins/pytorch/test.sh:
```diff
@@ -401,9 +412,21 @@ else
+    time lcov --capture --directory . --output-file coverage.info
+    # remove files that we do not care about from coverage report
+    time lcov --remove coverage.info '/usr/*' '*/third_party/*' '*/caffe2/*' --output-file coverage.info
+    # remove .AVX2.cpp suffix and remove build/ in filename in coverage.info
+    time python ../tools/code_coverage/package/tool/modify_coverage_info.py coverage.info
```
This better be done using `fixes:` section of https://github.com/pytorch/pytorch/blob/master/codecov.yml

scintiller replied:
I didn't find a way to remove `*.AVX2.cpp`. Besides, this script may also be used locally for `gcc` coverage report in the future.

scintiller committed:
Change code coverage option name (#43999)

Summary:
According to [documentation](https://github.com/pytorch/pytorch/blob/master/tools/setup_helpers/cmake.py#L265), only options starts with `BUILD_` / `USE_` / `CMAKE_` in `CMakeLists.txt` can be imported by environment variables.

 ---
This diff is originally intended to enable  `c++` source coverage with `CircleCI` and `codecov.io`, but we will finish it in the future. You can find the related information in the diff history. Following is the originally procedur:

Based on [this pull request](https://github.com/pytorch/pytorch/commit/1bda5e480c0fe01dae8db718cd457befa1b46ac6), life becomes much easier for this time.
1.in `build.sh`
- Enable coverage builld option for c++
- `apt-get install lcov`

2.in `test.sh`
- run `lcov`

3.in `pytorch-job-specs.yml`
- copy coverage.info to `test/` folder and upload it to codecov.io

Pull Request resolved: https://github.com/pytorch/pytorch/pull/43999

Test Plan: Test on github

Differential Revision: D23464656

Pulled By: scintiller

fbshipit-source-id: b9cdb35d4466147075026720cd993dd0c5093cad
